v-next/hardhat/src/internal/builtin-plugins/node/task-action.ts

```diff
@@ -129,7 +138,58 @@ const nodeAction: NewTaskActionFunction<NodeActionArguments> = async (
 
   console.log();
 
-  // TODO(https://github.com/NomicFoundation/hardhat/issues/6040): Add build info watcher here
+  const buildInfoDirPath = path.join(
+    hre.config.paths.artifacts,
+    BUILD_INFO_DIR_NAME,
+  );
+  const addCompilationResult = async (modifiedPath: string) => {
+    log(`Detected change in ${modifiedPath}`);
+
+    // NOTE: We're ignoring the output file here, because the build info files
+    // are modified after the output files
```

Comment:
Actually, we write to them concurrently. So we need to somehow detect when both build info file and its output have been written to.